### ISSUE TITLE
Estimating random utility models from Bruhin et al. (2019)

### ISSUE THREAD

**Issue by armoutihansen** [arteagac/xlogit#10]
First of all: Thanks a lot for this very useful package!

I was wondering whether it would be possible to estimate the random utility models in Bruhin et al. (2019) as mixed logits using xlogit? The authors already estimate the utility models as standard logits and finite mixtures. Specifically, they estimate simple social preference models based on a panel of binary dictator and reciprocity games. In a binary dictator game, the subject is randomly matched with another subject and has to choose between two allocations. The utility of a given allocation is given by:

<img width="1256" alt="image" src="https://user-images.githubusercontent.com/28649389/182459986-0d5f2de4-9352-4749-ba34-f10451058b22.png">

In a binary reciprocity game, the subject is faced with the same decision, but the matched subject performs either a kind or unkind decision beforehand. Hence, the utility here is given by:

<img width="1326" alt="Screenshot 2022-08-01 at 19 33 58" src="https://user-images.githubusercontent.com/28649389/182460373-7d221029-9a8f-4dd3-b81e-2cfb7ca57b36.png">

Based on this, the probability of choosing one allocation over the other is given by:

<img width="779" alt="Screenshot 2022-08-02 at 10 35 49" src="https://user-images.githubusercontent.com/28649389/182460667-9b39b846-11a7-4e95-8c5f-b91d2171f1b3.png">

It is not clear to me whether such a functional form, both as standard logit and mixed logit, can be estimated by xlogit. 

Many thanks and best,
Jesper

**Comment by arteagac** [comment]
Hi @armoutihansen! My understanding is that your goal is to estimate the parameters $\alpha$ and $\beta$. If this is the case, perhaps you can re frame the utility specification as shown below (the same can be done for the utility in Equation (2)). The utility specification below could be used in `xlogit` to estimate a Mixed Logit model and select random parameters for $\alpha$ and $\beta$. 

$U^A =  (1 - \alpha s - \beta r)\prod{^A}  + (\alpha s + \beta r )\prod{^B}   $
$U^A = \prod{^A} - \prod{^A}\alpha s - \prod{^A}\beta r + \prod{^B}\alpha s + \prod{^B}\beta r     $
$U^A = \prod{^A} + \alpha (-\prod{^A} s +  \prod{^B} s)  + \beta(- \prod{^A} r  + \prod{^B} r)  $

Let me know if this utility specification is valid, and I can further guide you on how to estimate it in `xlogit`.

**Comment by armoutihansen** [comment]
Hi @arteagac! Thanks a lot for the quick response.

That is indeed correct. The goal would be to estimate $(\alpha,\beta)$ in

$$U^A=\Pi^A+\alpha(\Pi^B s - \Pi^A s)+\beta((\Pi^B r - \Pi^A r)$$

in the first specification as well as $(\alpha,\beta,\gamma,\delta)$ in

$$U^A=\Pi^A+\alpha(\Pi^B s - \Pi^A s)+\beta((\Pi^B r - \Pi^A r) + \gamma((\Pi^B q - \Pi^A q) + \delta((\Pi^B v - \Pi^A v)$$

in the second. Furthermore, for both of these, the goal would also be to estimate the choice sensitivity/scale parameter $\sigma$ given in (4) above.

**Comment by arteagac** [comment]
Hi @armoutihansen. Great! It seems that `xlogit` can indeed help you with this estimation. You simply need to prepare a dataset that pre-computes $s(\prod{^B} - \prod{^A})$, $r(\prod{^B} - \prod{^A})$, $q(\prod{^B} - \prod{^A})$, and $v(\prod{^B} - \prod{^A})$, and then you can use these as input columns in`xlogit` to estimate the model ([see estimation examples here](https://xlogit.readthedocs.io/en/latest/notebooks/mixed_logit_model.html)). If the examples are not clear or if you need any further help, please do not hesitate to let me know
For the sensitivity/scale parameter $\sigma$, I think you can frame it as a WTP-type of problem, in which the willingness to pay is modeled using a scale factor (usually the negative of the price), which makes the problem non-linear. Fortunately, `xlogit` also supports this type of estimation. For this, you simply need to pass `scale_factor` as an argument to the fit function.

**Comment by armoutihansen** [comment]
Hi @arteagac. Thank you very much for the help! 

For the standard (multinomial) logit, I managed to reproduced the results of Bruhin et al. (2019):

<img width="1387" alt="image" src="https://user-images.githubusercontent.com/28649389/182596031-04fd6155-26a2-4183-a188-7af1faa85e48.png">

In the output above, $(x_1,\dots,x_4)=(s(\Pi^B-\Pi^A),\dots,v(\Pi^B-\Pi^A))$ as you suggested in your previous reply. However, the signs of $(x_1,\dots,x_4)$ are opposite that of Bruhin et al. I tried setting `scale_factor=-df_long['pi']` to add instead of subtract, but this returned different parameter estimates and a much lower logik. Do you know how to deal with this? Also: It is not (yet) possible to cluster the standard errors, right?

For the mixed logit, I did not manage to get that encouraging results:

<img width="1979" alt="image" src="https://user-images.githubusercontent.com/28649389/182598034-4453fd07-18a8-4c8c-ba81-8008fccd2b0a.png">

It is not clear to me why the loglik is so much lower than that of the multinomial logit. Also, if I omit the scale factor, I achieve almost half as low a loglik. I am wondering whether this is due to the scale factor not being random in the current estimation. I saw that this point was discussed in your WTP issue, but I couldn't find any reference to randomising the scale factor in the documentation. Is is correct that it is not (yet) possible to let the scale factor be random?

Again, thank you very much for the help!

**Comment by arteagac** [comment]
Hi @armoutihansen, I am not sure what might be causing the flip in signs compared to Bruhin et.al., but I assume it might be some minor issue when you pre-processed x1, x2, x3, and x4. Make sure you did not flip any operations during the data preparation. Regarding clustered errors, unfortunately xlogit still does not support those.

Regarding your mixed logit model, I think there is something off because the log-likelihood is extremely low. This is a potential symptom of non-convergence, which is non-surprising for this type of non-linear WTP-like models. You are right, xlogit still does not support a random scale factor, but I think this might not be the cause of the issue. I would advise to try running multiple estimations using different start points by passing the to the `init_coeff` argument in the fit function. Also, if for your case it is critical to use a random scale_factor, perhaps you can take a look at the logitr R package (https://jhelvy.github.io/logitr/), which supports this feature.

**Comment by armoutihansen** [comment]
Okay. Thanks a lot! I will give logitr a try for the clustered standard errors and random scale factor.

**Comment by arteagac** [comment]
Sure, please let me know how it goes.

**Comment by armoutihansen** [comment]
Hi @arteagac, I just wanted to quickly update you on my progress.

First of all, I realised that I do not need to specify my model in WTP space in order to estimate the scale parameter. In particular, I can just estimate:

$$\frac{1}{\sigma}\Pi^A +\frac{\alpha}{\sigma}(s\Pi^B-s\Pi^A)+\frac{\beta}{\sigma}(r\Pi^B-r\Pi^A)+\frac{\gamma}{\sigma}(q\Pi^B-q\Pi^A)+\frac{\delta}{\sigma}(v\Pi^B-v\Pi^A)$$

and then multiply $\alpha,\beta,\gamma,\delta$ by $\sigma$ after estimation. Naturally, $\frac{1}{\sigma}$ should be positive, so I specify a truncated normal (I assume that this is identical to the zero censured normal in logitr. Correct me if I am wrong) for that and a normal distribution for the other parameters. Based on this, I was able to achieve convergence in logitr, but not xlogit. 

logitr:
<img width="608" alt="image" src="https://user-images.githubusercontent.com/28649389/183888668-b62faa44-6b63-42f4-8acd-f2a725a80138.png">
Log-likelihood: -3,819.24

xlogit:
<img width="1288" alt="image" src="https://user-images.githubusercontent.com/28649389/183889226-f45f365f-5fd6-416c-b40b-67fa1878913e.png">

In the outputs above, `self`is $\frac{1}{\sigma}$, so it is not clear to me why it is negative when I specify `tn`, but perhaps I have misunderstood how the truncated normal is used? If I specify it as log-normally distributed, I am far from convergence with a log-likelihood of approximately -50,000.

**Comment by arteagac** [comment]
Hi @armoutihansen, Thanks a lot for the update. Your feedback helps me to keep improving xlogit. I took a look at logitr's censored normal implementation and it seems that it is the same as xlogit's truncated normal (I will double-check if it is better to rename `tn` as `cn` in xlogit to keep consistency with logitr). Technically, both should provide the same results, but I am not sure why xlogit fails to converge. I know logitr uses by default the `L-BFGS-B` optimization routine, whereas xlogit uses by default `BFGS`. Perhaps you can try setting xlogit to use `L-BFGS-B`  by using the `optim_method='L-BFGS-B'` parameter in the `fit` method. `L-BFGS-B` is more robust, so perhaps this can help with the convergence issue.
